### DISCUSSION TITLE
x-for loop with async capabilities

### DISCUSSION THREAD

**Discussion by Jamiul-Bari** [alpinejs/alpine#4058] [1. Help] [answered]
I am using Livewire and Alpine for a Laravel project. In a particular component, I try to search and filter locations (say more than 20k locations), but it freezes as soon as I type something. So, I want to do it asynchronously. Is there any way to use x-for asynchronously?

```
<div
    class="z-10 w-full rounded-lg shadow divide-y" x-data="{
    selectedArea: {
        area_id: 0,
        area_name: ''
    },
    areas: [],
    getAreas: async function() {
        fetch(`${$wire.url}/api/all-areas`)
            .then((response) => response.json())
            .then((json) => this.areas = json);
    },
    filteredAreas: async function(onlyArea) {
        const re = RegExp(`.*${this.searchLocation.toLowerCase().split('').join('.*')}.*`);

         return this.areas.filter(v => v.area_name.toLowerCase().match(re));
    },
}">
    
   <input x-model='searchLocation' type="text" class="w-full" autocomplete="off" />

    <ul class="py-2 text-sm text-gray-200 location-search-result h-80 overflow-y-scroll scrollbar">
        <template x-for="option in await filteredAreas(false)">
            <li>
                <a @click.prevent="selectedArea = option"
                    class="block px-4 py-2 hover:bg-primary hover:text-white break-all"
                    x-text="option.area_name">
                </a>
            </li>
        </template>
    </ul>
</div>
```

I tried `async await` but it still freezes. Any suggestions?

**Comment by ekwoka** [comment, marked as answer]
It's all synchronous operations that you're doing. So there is no way to do it "async". Here you aren't actually doing anything "async" aside from the fetch.

It's probably smart for you to paginate the results.

Or if you just can't paginate/limit, you can incrementally increase the number rendered with some kind of a limit over time.

The part that makes it slow is the actual creation of the elements, so the only way to make it faster is to create less elements.

since alpine will reuse existing elements that match, you can just slowing increase the number rendered and Alpine will MOSTLY only need to do work on the new items, not the old.

**Reply by Jamiul-Bari** [reply]
Your approach makes more sense. I thought using async would create a new thread and leave the main thread smooth, like Java GUI applications. Thank you for sharing your knowledge. Now, I am accepting a query parameter in the API, filtering the results from the backend, and sending paginated results.

**Reply by ekwoka** [reply]
async is not multi threaded/parallel

It's just asynchronous.

It's just concurrent. That they run out of order.

But here, all that means is that the expression runs, runs the function that filters the list, and then puts the rendering of the elements onto the microtask queue, which then processes synchronously as well, never allowing the main thread time to paint the layout.
